pkg/pillar/cmd/downloader/downloader.go

```diff
@@ -1176,6 +1183,166 @@ func doSftp(ctx *downloaderContext, status *types.DownloaderStatus,
 	}
 }
 
+func rktFetch(url string, localConfigDir string) (string, error) {
+	// rkt fetch --system-config=/persist/rkt-local/ --insecure-options=image
+	// rkt --insecure-options=image fetch docker://zededa/zcli --full=true
+	//      docker://zededa/zcli-dev:latest
+	log.Debugf("rktFetch - url: %s ,  localConfigDir:%s\n",
+		url, localConfigDir)
+	cmd := "rkt"
+	args := []string{
+		"--dir=" + persistRktDataDir,
+		"--insecure-options=image",
+		"fetch",
+	}
+	// if len(localConfigDir) > 0 {
+	// 	args = append(args, "--system-config="+persistRktLocalConfigDir)
+	// }
+	args = append(args, url)
+	args = append(args, "--full=true")
+
+	log.Infof("rktFetch - url: %s ,  localConfigDir:%s, args: %+v\n",
+		url, localConfigDir, args)
+
+	stdoutStderr, err := wrap.Command(cmd, args...).CombinedOutput()
+	if err != nil {
+		log.Errorln("rkt fetch failed ", err)
+		log.Errorln("rkt fetch output ", string(stdoutStderr))
+		return "", fmt.Errorf("rkt fetch failed: %s\n",
+			string(stdoutStderr))
+	}
+	log.Infof("rktFetch - image fetch successful. stdoutStderr: %s\n",
+		stdoutStderr)
+	outputStr := string(stdoutStderr)
+	log.Debugf("rktFetch - outputStr: %s\n", outputStr)
+	outputStrArray := strings.Split(outputStr, "\n")
+
+	log.Debugf("rktFetch - outputStrArray:\n")
+	for i, op := range outputStrArray {
+		log.Debugf("index:%d, op:%s", i, op)
+	}
+	log.Debugf("rktFetch - outputStrArray DONE\n")
+
+	// Get ImageID from the oputput. The last line in rkt fetch output
+	// with sha12- is the imageID
+	imageID := ""
+	for i := len(outputStrArray) - 1; i >= 0; i-- {
+		imageID = outputStrArray[i]
+		if strings.HasPrefix(imageID, "sha512-") {
+			break
+		}
+	}
+	log.Infof("rktFetch - imageID: %s\n", imageID)
+	if imageID == "" {
+		errMsg := "rkt fetch: Can't find imageID.\n Fetch Output: " +
+			outputStr
+		return "", errors.New(errMsg)
+	}
+
+	// XXX:FIXME - we should run "rkt image ls" and verify image fetch
+	// went thru without errors.
+	return imageID, nil
+}
+
+func rktAuthFilename(appName string) string {
+	return persistRktLocalConfigAuthDir + "/rktAuth" + appName + ".json"
+}
+
+func rktCreateAuthFile(config *types.DownloaderConfig) (string, error) {
+
+	if len(strings.TrimSpace(config.ApiKey)) == 0 {
+		log.Debugf("rktCreateAuthFile: empty username. Skipping AuthFile")
+		return "", nil
+	}
+
+	// XXX:FIXME - Why should this directory be created again
```

Comment:
Please answer this question.

Reply:
Removed the comment.. We are creating the dir. here instead of in device steps. Downloader is the agent that needs this directory - so we will keep it here.